### DISCUSSION TITLE
When function `stream_socket_server` disabled, error message not clear

### DISCUSSION THREAD

**Discussion by xxm404** [reactphp/reactphp#528] [Q&A]
When i use phpstan in my php project, i get error message like this
```
stream_set_blocking() expects parameter 1 to be resource, null given in phar:///home/work/env/php/bin/phpstan/vendor/react/socket/src/TcpServer.php on line 165
```
from TcpServer.php source code i find this code https://github.com/reactphp/socket/blob/1.x/src/TcpServer.php#L170
```
$this->master = @\stream_socket_server(
```
in my local `php.ini` `stream_socket_server` function is disabled, so the `$this->master` will be `NULL`, `stream_socket_server` function be dissabled error message is suppressed

Why use "@" when calling stream_socket_server function? This introduces some additional cost to positioning errors

**Comment by WyriHaximus** [comment]
Thank you for reporting this, when using this package we're assuming you have don't have any required function disabled so we don't check for that. We do check if the call has failed but apparently, the return value is different when said function has been disabled: 

https://github.com/reactphp/socket/blob/1.x/src/TcpServer.php#L170-L188

As you can see in the snipped we capture the error in a different way to throw an exception instead of letting it go to STDERR.

**Reply by xxm404** [reply]
@WyriHaximus yes i got you, when `stream_socket_server ` is disabled, the return value is `NULL` not `false`
Maybe we can check the `stream_socket_server` function like this
```
$disabled = explode(',', ini_get('disable_functions'));
$is_disable = in_array('stream_socket_server', $disabled);
```

**Reply by WyriHaximus** [reply]
That would mean doing an additional check before calling it, while we also get the result by attempting to call it, which would be cheaper.

Another thing is that this package expects you to allow these functions, hence we're not checking for them to be enabled. Not sure if it's our responsibility to check if we can use the function before using it. We could improve the error handling around it.

**Reply by xxm404** [reply]
> Another thing is that this package expects you to allow these functions, hence we're not checking for them to be enabled. Not sure if it's our responsibility to check if we can use the function before using it. 

Yes you can expect this function to be allowed if someone uses this package directly, but maybe there are some scenarios where this package is used in some third-party applications(e.g. phpstan), a clear error message help users.

**Reply by SimonFrings** [reply]
@xxm404 PHPStan is using ReactPHP under the hood and seems to need the `stream_socket_server` function, so disabling this results in said error message. I agree though that we need a clearer error message for this, but this would still result in PHPStan reporting an error. I think it makes sense to open a ticket for this in https://github.com/phpstan/phpstan.

Additionally we're always happy about PRs to give this a better error message, this could be achieved by checking with `function_exists` if `stream_socket_server` is disabled. Changing this should also include a test where we disable the `stream_socket_server` function and then check for the right error message.

Hope this helps :+1:
